### PR TITLE
feat: dispatch permissions to frontend in experimental contexts mode

### LINKED ISSUE #7629: Use equivalent of `kubectl auth can-i` on Kube contexts to check connectivity instead of informers
### Is your enhancement related to a problem? Please describe

The current implementation for checking the connectivity to Kubernetes contexts uses informers. This solution has some drawbacks:
- the informer is not able to detect that the cluster is unreachable; Podman Desktop needs to define a timeout after which the context is considered unreachable, which complicates and makes the result less accurate
- Podman Desktop doesn't know if the user has the credentials to watch the resources, but we are always watching all resources. The user having limited credentials may result in wrong information (for example for nodes, which are generally not readable in shared clusters, for example OpenShift Sandbox)

### Describe the solution you'd like

I propose to query the [`SelfSubjectRulesReview`](https://kubernetes.io/docs/reference/kubernetes-api/authorization-resources/self-subject-rules-review-v1/) resource in the cluster to check the connectivity to the cluster, before to start informers.
The advantages are:
- the request is using a POST HTTP operation, with which Podman Desktop can detect connectivity error (still using some timeout)
- the `SelfSubjectRulesReview` are expected to be accessible for all users
- Podman Desktop can use the information to start informers only on the resources the user can watch. Ultimately, we could also check the `list` permission, to have a degraded implementation using `list` operation and not `watch` (informers), if the user has `list` but not `watch` permissions.

### TODO

- [x] backend part: #10133  
- [x] dispatch to frontend: #10383
- [ ] frontend part: #10655

### PR DESCRIPTION
**Pull request by feloy**
### What does this PR do?

- Dispatch permissions to frontend in experimental contexts mode
- the latest commit refactors existing code to use `KubernetesResourceName` instead of `string` 

### Screenshot / video of UI

No UI changes in this PR

UI will come after https://github.com/podman-desktop/podman-desktop/issues/10385

### What issues does this PR fix or reference?

Fixes #7629

### How to test this PR?

- [x] Tests are covering the bug fix or the new feature

### CODE REVIEW

File: packages/api/src/kubernetes-resources.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+export const kubernetesResources = [
```
Comment (benoitf) [inline packages/api/src/kubernetes-resources.ts]: issue: looks odd to have a constant with camelCase pattern. Should be all UPPER_CASE

File: packages/main/src/plugin/kubernetes/contexts-states-dispatcher.ts [hunk 2]
```diff
@@ -56,6 +57,21 @@ export class ContextsStatesDispatcher {
...
+      for (const [resourceName, contextResourcePermission] of permissions) {
+        value.push({
+          contextName,
+          resourceName,
+          permitted: contextResourcePermission.permitted,
+          reason: contextResourcePermission.reason,
+        });
+      }
+    }
+    return value;
```
Comment (benoitf) [inline packages/main/src/plugin/kubernetes/contexts-states-dispatcher.ts]: question: should it use the .map pattern rather two for loop ?

File: packages/main/src/plugin/kubernetes/kubernetes-client.ts [hunk 3]
```diff
@@ -1852,4 +1853,11 @@ export class KubernetesClient {
     }
     return this.contextsStatesDispatcher?.getContextsHealths();
   }
+
+  public getContextsPermissions(): ContextPermission[] {
+    if (!this.contextsStatesDispatcher) {
+      throw new Error('contextsStatesDispatcher is undefined. This should not happen in Kubernetes experimental');
+    }
+    return this.contextsStatesDispatcher?.getContextsPermissions();
```
Comment (benoitf) [inline packages/main/src/plugin/kubernetes/kubernetes-client.ts]: TypeScript is not able to infer that it's always here ?
```suggestion
    return this.contextsStatesDispatcher.getContextsPermissions();
```

**Comment by feloy** [comment]
PR set in draft, as I'm not sure if it is appropriate to have a strict type KubernetesResourceName for resources names, instead of a generic `string` type, as this could block us from managing CRDs. 

I'll make a quick PoC to check how CRDs could be handled, and if we may use or not a strict type for resources names.

=> #10418

With this PoC, we cannot have a strict type for resource names

**Review by benoitf — COMMENTED** [review]
looks like I forgot to hit send when I read the PR...

File: packages/api/src/kubernetes-resources.ts (hunk 1 above)
Comment (feloy) [reply packages/api/src/kubernetes-resources.ts]: I finally removed this type

File: packages/api/src/kubernetes-contexts-permissions.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  resourceName: string;
+  // permitted if allowed and not denied
+  // > When multiple authorization modules are configured, each is checked in sequence.
+  // > If any authorizer approves or denies a request, that decision is immediately returned
+  // > and no other authorizer is consulted. If all modules have no opinion on the request,
+  // > then the request is denied. An overall deny verdict means that the API server rejects
+  // > the request and responds with an HTTP 403 (Forbidden) status.
+  // (source: https://kubernetes.io/docs/reference/access-authn-authz/authorization/)
+  permitted: boolean;
+  reason?: string;
```
Comment (benoitf) [inline packages/api/src/kubernetes-contexts-permissions.ts:34]: nit: document this optional attribute ?

**Review by benoitf — APPROVED** [review]

File: packages/renderer/src/stores/kubernetes-context-permission-experimental.spec.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  value: {
+    kubernetesGetContextsPermissions: vi.fn(),
+    getConfigurationValue: vi.fn(),
+    addEventListener: eventEmitter.receive,
+    events: {
+      receive: eventEmitter.receive,
+    },
+  },
+  writable: true,
+});
```
Comment (benoitf) [inline packages/renderer/src/stores/kubernetes-context-permission-experimental.spec.ts]: question: should it be in the beforeAll method ?

**Review by benoitf — APPROVED** [review]

File: packages/renderer/src/stores/kubernetes-context-permission-experimental.spec.ts (hunk 5 above)
Comment (feloy) [reply packages/renderer/src/stores/kubernetes-context-permission-experimental.spec.ts]: yes, that seems more reasonable. Moving it

File: packages/renderer/src/stores/kubernetes-context-permission-experimental.spec.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  Object.defineProperty(global, 'window', {
+    value: {
+      kubernetesGetContextsPermissions: vi.fn(),
+      getConfigurationValue: vi.fn(),
+      addEventListener: eventEmitter.receive,
+      events: {
+        receive: eventEmitter.receive,
+      },
+    },
+    writable: true,
```
Comment (jeffmaury) [inline packages/renderer/src/stores/kubernetes-context-permission-experimental.spec.ts]: Is that required ?

File: packages/renderer/src/stores/kubernetes-context-permission-non-experimental.spec.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  Object.defineProperty(global, 'window', {
+    value: {
+      kubernetesGetContextsPermissions: vi.fn(),
+      getConfigurationValue: vi.fn(),
+      addEventListener: eventEmitter.receive,
+      events: {
+        receive: eventEmitter.receive,
+      },
+    },
+    writable: true,
```
Comment (jeffmaury) [inline packages/renderer/src/stores/kubernetes-context-permission-non-experimental.spec.ts]: Same here ?

File: packages/renderer/src/stores/kubernetes-context-permission-experimental.spec.ts (hunk 6 above)
Comment (feloy) [reply packages/renderer/src/stores/kubernetes-context-permission-experimental.spec.ts]: removed

**Review by jeffmaury — APPROVED** [review]
LGTM

**Pull request merged**